File: pkg/yeelight_property.py
"""Yeelight adapter for Mozilla WebThings Gateway."""

from gateway_addon import Property
from yeelight import BulbException


class YeelightProperty(Property):
    """Yeelight property type."""

    def __init__(self, device, name, description, value):
        """
        Initialize the object.

        device -- the Device this property belongs to
        name -- name of the property
        description -- description of the property, as a dictionary
        value -- current value of this property
        """
        Property.__init__(self, device, name, description)
        self.set_cached_value(value)

    def set_value(self, value):
        """
        Set the current value of the property.

        value -- the value to set
        """
        color_mode_prop = None
        if 'colorMode' in self.device.properties:
            color_mode_prop = self.device.properties['colorMode']

        try:
            self.device.update_properties()

            on = self.device.is_on()

            if self.name == 'on':
                if value:
                    self.device.bulb.turn_on()
                else:
                    self.device.bulb.turn_off()
            elif on:
                # only change the other properties if the bulb is already on
                if self.name == 'color':
                    self.device.bulb.set_rgb(int(value[1:3], 16),
                                             int(value[3:5], 16),
                                             int(value[5:7], 16))

                    # update the colorMode property
                    if color_mode_prop is not None:
                        color_mode_prop.set_cached_value('color')
                        self.device.notify_property_changed(color_mode_prop)
                elif self.name == 'level':
                    self.device.bulb.set_brightness(value)
                elif self.name == 'colorTemperature':
                    value = max(value, self.description['minimum'])
                    value = min(value, self.description['maximum'])
                    self.device.bulb.set_color_temp(value)

                    # update the colorMode property
                    if color_mode_prop is not None:
                        color_mode_prop.set_cached_value('temperature')
                        self.device.notify_property_changed(color_mode_prop)
                else:
                    return
        except BulbException:
            return

        self.set_cached_value(value)
        self.device.notify_property_changed(self)

    def update(self):
        """Update the current value, if necessary."""
        if self.name == 'on':
            value = self.device.is_on()
        elif self.name == 'color':
            value = self.device.color()
        elif self.name == 'level':
            value = self.device.brightness()
        elif self.name == 'colorTemperature':
            value = self.device.color_temp()
        elif self.name == 'colorMode':
            value = self.device.color_mode()
        else:
            return

        if value != self.value:
            self.set_cached_value(value)
            self.device.notify_property_changed(self)

File: lib/yeelight-property.js
'use strict';

const {Property} = require('gateway-addon');

class YeelightProperty extends Property {
  constructor(device, name, description, value) {
    super(device, name, description);
    this.setCachedValue(value);
  }

  setValue(value) {
    const colorModeProp = this.device.properties.get('colorMode');

    const on = this.device.on();

    if (this.name === 'on') {
      return this.device.light.set_power(value ? 'on' : 'off')
        .then(() => {
          this.setCachedValueAndNotify(value);
          return value;
        });
    }

    // only change the other properties if the bulb is already on
    if (!on) {
      return Promise.reject('cannot set property when light is off');
    }

    if (this.name === 'color') {
      const rgb = value.substring(1).match(/../g).map((x) => parseInt(x, 16));
      return this.device.light.set_rgb(rgb)
        .then(() => {
          // update the colorMode property
          if (colorModeProp) {
            colorModeProp.setCachedValueAndNotify('color');
          }

          this.setCachedValueAndNotify(value);
          return value;
        });
    }

    if (this.name === 'level') {
      return this.device.light.set_bright(value)
        .then(() => {
          this.setCachedValueAndNotify(value);
          return value;
        });
    }

    if (this.name === 'colorTemperature') {
      value = Math.max(value, this.minimum);
      value = Math.min(value, this.maximum);

      return this.device.light.set_ct_abx(value)
        .then(() => {
          // update the colorMode property
          if (colorModeProp) {
            colorModeProp.setCachedValueAndNotify('temperature');
          }

          this.setCachedValueAndNotify(value);
          return value;
        });
    }

    return Promise.reject('unknown property');
  }

  /**
   * Update the current value, if necessary.
   */
  update() {
    let value;
    switch (this.name) {
      case 'on':
        value = this.device.on();
        break;
      case 'color':
        value = this.device.color();
        break;
      case 'colorTemperature':
        value = this.device.colorTemp();
        break;
      case 'colorMode':
        value = this.device.colorMode();
        break;
      case 'level':
        value = this.device.brightness();
        break;
    }

    if (value !== this.value) {
      this.setCachedValueAndNotify(value);
    }
  }
}

module.exports = YeelightProperty;
